codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,297 @@
+    }
+
+    class FrameListener extends Http2FrameAdapter {
+        @Override
+        public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorCode)
+                throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().requireStream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+            // Use a user event in order to circumvent read queue.
+            streamInfo.childChannel.pipeline().fireUserEventTriggered(
+                new Http2ResetMessage(errorCode).setStreamId(streamId));
+        }
+
+        @Override
+        public void onHeadersRead(ChannelHandlerContext ctx, int streamId,
+                Http2Headers headers, int streamDependency, short weight, boolean
+                exclusive, int padding, boolean endStream) throws Http2Exception {
+            onHeadersRead(ctx, streamId, headers, padding, endStream);
+        }
+
+        @Override
+        public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
+                int padding, boolean endOfStream) throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().requireStream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+            fireChildReadAndRegister(streamInfo,
+                new Http2HeadersMessage(headers, endOfStream, padding).setStreamId(streamId));
+        }
+
+        @Override
+        public int onDataRead(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
+                boolean endOfStream) throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().requireStream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+            // TODO: Determine whether retain is needed.
+            fireChildReadAndRegister(streamInfo,
+                new Http2DataMessage(data, endOfStream, padding).setStreamId(streamId));
+            return 0;
+        }
+    }
+
+    static class StreamInfo {
+        final Http2StreamChannel childChannel;
+        /**
+         * True if stream is in {@link Http2MultiplexCodec#steamsToFireChildReadComplete}.
+         */
+        boolean inStreamsToFireChildReadComplete;
+
+        StreamInfo(Http2StreamChannel childChannel) {
+            this.childChannel = childChannel;
+        }
+    }
+
+    class Http2StreamChannel extends ChildChannel {
+        private final Http2Stream stream;
+
+        Http2StreamChannel(Http2Stream stream) {
+            super(Http2MultiplexCodec.this.ctx.channel());
+            this.stream = stream;
+        }
+
+        @Override
+        protected void doClose() throws Exception {
+            Http2StreamMessage resetMessage =
+                    new Http2ResetMessage(Http2Error.CANCEL).setStreamId(stream.id());
+            ctx.invoker().invokeWrite(ctx, resetMessage, ctx.newPromise());
```

Comment:
I'm guessing `ctx.writeAndFlush(..)` can't be used because it needs to be done now (before the close operation completes) or some state has already been updated which would prevent the write/flush? Add a comment?